### DISCUSSION TITLE
Invalid `client.user.update()` invocation error

### DISCUSSION THREAD

**Discussion by donghhan** [prisma/prisma#11843] [Q&A]
I am trying to make mutation called `editProfile` that allows users to edit their `firstName`, `email` and `password`. When they change the `password`, the changed `password` also encrypted via `bcrypt` npm package. 

I am using GraphQL and Prisma with Nexus

```typescript
// EditProfile Mutation
export const EditProfileMutation = mutationField("editProfile", {
  type: "OkResult",
  args: {
    firstName: stringArg(),
    username: stringArg(),
    email: stringArg(),
    password: stringArg(),
  },
  async resolve(_, args, ctx) {
    const { firstName, username, email, password: newPassword } = args;
    const { signedInUser } = ctx;

    let hashedPassword = null;

    if (newPassword) {
      hashedPassword = await bcrypt.hash(newPassword, 10);
    }

    const updateComplete = await client.user.update({
      where: { id: signedInUser?.id },
      data: {
        firstName,
        username,
        email,
        ...(hashedPassword && { password: hashedPassword }),
      },
    });

    if (updateComplete.id) {
      return {
        ok: true,
      };
    } else {
      return {
        error: "Could not update the profile.",
      };
    }
  },
});
```

And when I tried anything to change it in GraphQL playground, it threw error:

```
Invalid `client.user.update()` invocation in /Users/apensia/dev/fleam-api/src/graphql/User.ts:170:46 167 hashedPassword = await bcrypt.hash(newPassword, 10); 168 } 169 → 170 const updateComplete = await client.user.update({ where: { ? id?: String, ? username?: String, ? email?: String }, data: { firstName: 'Donghoon', username: undefined, email: undefined } }) Argument where of type UserWhereUniqueInput needs at least one argument. Available args are listed in green. Note: Lines with ? are optional.
```

As I can tell from its error, I think something went wrong with `bcrypt`. However, I just don't understand why it keeps complaining about it because I surely connected reducer function with `async` and `await`. 

Can anyone help on this?

**Comment by tarazena** [comment]
your issue is in `...(hashedPassword && { password: hashedPassword }),`, try 
```ts
const updateComplete = await client.user.update({
      where: { id: signedInUser?.id },
      data: {
        firstName,
        username,
        email,
        ...(hashedPassword ? { password: hashedPassword } : {}),
      },
    });
```

**Reply by donghhan** [reply]
It did not work, keep creating the same error. 

I checked whether my JWT token properly enters into my headers. But it keeps failing the login.

**Reply by tarazena** [reply]
Oh I see, If im not wrong, your schema are asking username and email to be not nullable, in your update you are passing undefined for email and username, which its causing prisma to throw an exception.

**Reply by vladi-stevanovic** [reply]
👋 Hello @donghhan  I'm going through pending GitHub Discussions to check if you need any help. Is this issue still open?
Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! 🙌

**Comment by theonlineaid** [comment]
### I think this code will help you  here I use [zod](https://zod.dev/) for data validation

```
export async function POST(request: NextRequest) {
  const body = await request.json();
  const validation = schema.safeParse(body);

  if (!validation.success)
    return NextResponse.json(validation.error.errors, { status: 400 });

  const user = await prisma.user.findUnique({
    where: { email: body.email },
  });

  if (user)
    return NextResponse.json({ error: "User already exists" }, { status: 400 });

  const hashedPassword = await bcrypt.hash(body.password, 10);

  const newUser = await prisma.user.create({
    data: {
      email: body.email,
      password: hashedPassword
      // hashedPassword,   // don't use this line
      // ...(hashedPassword ? { password: hashedPassword } : {}),  // don't use this line
    },
  });

  return NextResponse.json({ email: newUser.email }, { status: 201 });
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
